From 56ae25b3a5ec0ca284325a97ca80b2d68c2f65d0 Mon Sep 17 00:00:00 2001
From: Meng Yi <meng.yi@nxp.com>
Date: Sun, 18 Sep 2016 11:54:58 +0800
Subject: [PATCH 176/508] clocksource: Add timeout for while loop within IRQ
 context

Polling hardware in interrupt context within while loop is not
reasonable. Add a sensible timeout to abort this loop in case
of defective hardware

Signed-off-by: Meng Yi <meng.yi@nxp.com>
[Original patch taken from SDK-V2.0-1703]
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/clocksource/fsl_ftm_timer.c    |    4 +++-
 drivers/soc/fsl/layerscape/ftm_alarm.c |    4 +++-
 2 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/drivers/clocksource/fsl_ftm_timer.c b/drivers/clocksource/fsl_ftm_timer.c
index 900272d..6fc3637 100644
--- a/drivers/clocksource/fsl_ftm_timer.c
+++ b/drivers/clocksource/fsl_ftm_timer.c
@@ -83,7 +83,9 @@ static inline void ftm_counter_disable(void __iomem *base)
 
 static inline void ftm_irq_acknowledge(void __iomem *base)
 {
-	while (FTM_SC_TOF & ftm_readl(base + FTM_SC))
+	unsigned long timeout = jiffies + msecs_to_jiffies(100);
+
+	while ((FTM_SC_TOF & ftm_readl(base + FTM_SC)) && time_before(jiffies, timeout))
 		ftm_writel(ftm_readl(base + FTM_SC) & (~FTM_SC_TOF), base + FTM_SC);
 }
 
diff --git a/drivers/soc/fsl/layerscape/ftm_alarm.c b/drivers/soc/fsl/layerscape/ftm_alarm.c
index 1cc5891..581468c 100644
--- a/drivers/soc/fsl/layerscape/ftm_alarm.c
+++ b/drivers/soc/fsl/layerscape/ftm_alarm.c
@@ -78,7 +78,9 @@ static inline void ftm_counter_disable(void __iomem *base)
 
 static inline void ftm_irq_acknowledge(void __iomem *base)
 {
-	while (FTM_SC_TOF & ftm_readl(base + FTM_SC))
+	unsigned long timeout = jiffies + msecs_to_jiffies(100);
+
+	while ((FTM_SC_TOF & ftm_readl(base + FTM_SC)) && time_before(jiffies, timeout))
 		ftm_writel(ftm_readl(base + FTM_SC) & (~FTM_SC_TOF), base + FTM_SC);
 }
 
-- 
1.7.5.4

